File: apps/nongyu-go-track-server/internal/ingest/validate.go
package ingest

import (
	"encoding/json"
	"fmt"
	"strings"
	"unicode/utf8"
)

const maxPropsBytes = 4096

var allowedTypes = map[string]struct{}{
	"screen_view":    {},
	"button_click":   {},
	"perf":           {},
	"app_open":       {},
	"heartbeat":      {},
	"crash":          {},
	"llm_proxy_fail": {},
}

type RawEvent struct {
	EventID     string          `json:"event_id"`
	EventType   string          `json:"event_type"`
	EventName   string          `json:"event_name"`
	ClientTsMs  *int64          `json:"client_ts_ms"`
	SessionID   string          `json:"session_id"`
	AppVersion  string          `json:"app_version"`
	Platform    string          `json:"platform"`
	DeviceBrand string          `json:"device_brand"`
	DurationMs  *int64          `json:"duration_ms"`
	StudentNo   string          `json:"student_no"`
	Props       json.RawMessage `json:"props"`
}

type ItemError struct {
	EventID string `json:"event_id"`
	Code    string `json:"code"`
	Message string `json:"message,omitempty"`
}

type Prepared struct {
	Row        EventFields
	SkipInsert bool
	Err        *ItemError
}

type EventFields struct {
	EventID     string
	EventType   string
	EventName   string
	AppVersion  string
	Platform    string
	DeviceBrand string
	SessionID   string
	DurationMs  *int64
	PropsJSON   string
	ClientTsMs  *int64
	StudentNo   string
}

// ValidateOne 校验单条事件；失败时返回 ItemError，不阻断整包。
func ValidateOne(raw RawEvent) (*EventFields, *ItemError) {
	id := strings.TrimSpace(raw.EventID)
	if id == "" {
		return nil, &ItemError{EventID: raw.EventID, Code: "INVALID_EVENT", Message: "event_id required"}
	}
	typ := strings.TrimSpace(raw.EventType)
	if _, ok := allowedTypes[typ]; !ok {
		return nil, &ItemError{EventID: id, Code: "INVALID_TYPE", Message: "unknown event_type"}
	}
	name := strings.TrimSpace(raw.EventName)
	if name == "" {
		return nil, &ItemError{EventID: id, Code: "INVALID_EVENT", Message: "event_name required"}
	}
	platform := strings.TrimSpace(raw.Platform)
	if platform != "" && platform != "ios" && platform != "android" {
		return nil, &ItemError{EventID: id, Code: "INVALID_EVENT", Message: "invalid platform"}
	}

	propsJSON, err := normalizeProps(raw.Props)
	if err != nil {
		return nil, &ItemError{EventID: id, Code: "INVALID_EVENT", Message: "invalid props"}
	}

	return &EventFields{
		EventID:     id,
		EventType:   typ,
		EventName:   name,
		AppVersion:  strings.TrimSpace(raw.AppVersion),
		Platform:    platform,
		DeviceBrand: strings.TrimSpace(raw.DeviceBrand),
		SessionID:   strings.TrimSpace(raw.SessionID),
		DurationMs:  raw.DurationMs,
		PropsJSON:   propsJSON,
		ClientTsMs:  raw.ClientTsMs,
		StudentNo:   strings.TrimSpace(raw.StudentNo),
	}, nil
}

func normalizeProps(raw json.RawMessage) (string, error) {
	if len(raw) == 0 || string(raw) == "null" {
		return "", nil
	}
	var obj map[string]any
	if err := json.Unmarshal(raw, &obj); err != nil {
		return "", err
	}
	b, err := json.Marshal(obj)
	if err != nil {
		return "", err
	}
	if len(b) <= maxPropsBytes {
		return string(b), nil
	}
	trimmed := map[string]any{"_truncated": true, "_original_bytes": len(b)}
	if msg, ok := obj["message"]; ok {
		trimmed["message"] = clipAny(msg, 512)
	}
	if st, ok := obj["stack"]; ok {
		trimmed["stack"] = clipAny(st, 2048)
	}
	out, err := json.Marshal(trimmed)
	if err != nil {
		return "", err
	}
	return string(out), nil
}

func clipAny(v any, max int) string {
	s := fmt.Sprint(v)
	if utf8.RuneCountInString(s) <= max {
		return s
	}
	runes := []rune(s)
	return string(runes[:max])
}

File: apps/nongyu-node-track-server/src/ingest/validate.ts
import { TRACK_SERVER_EVENT_TYPES } from "nongyu-track-contract";

const maxPropsBytes = 4096;
const allowedTypes = new Set<string>(TRACK_SERVER_EVENT_TYPES);

export type RawEvent = {
  event_id?: string;
  event_type?: string;
  event_name?: string;
  client_ts_ms?: number | null;
  session_id?: string;
  app_version?: string;
  platform?: string;
  device_brand?: string;
  duration_ms?: number | null;
  student_no?: string;
  props?: unknown;
};

export type ItemError = {
  event_id: string;
  code: string;
  message?: string;
};

export type EventFields = {
  eventId: string;
  eventType: string;
  eventName: string;
  appVersion: string;
  platform: string;
  deviceBrand: string;
  sessionId: string;
  durationMs: number | null;
  propsJson: string;
  clientTsMs: number | null;
  studentNo: string;
};

function clipAny(v: unknown, max: number): string {
  const s = String(v);
  if ([...s].length <= max) return s;
  return [...s].slice(0, max).join("");
}

function normalizeProps(raw: unknown): string {
  if (raw === undefined || raw === null) return "";
  if (typeof raw !== "object" || Array.isArray(raw)) {
    throw new Error("invalid props");
  }
  const obj = raw as Record<string, unknown>;
  const b = JSON.stringify(obj);
  if (Buffer.byteLength(b, "utf8") <= maxPropsBytes) return b;
  const trimmed: Record<string, unknown> = {
    _truncated: true,
    _original_bytes: Buffer.byteLength(b, "utf8"),
  };
  if ("message" in obj) trimmed.message = clipAny(obj.message, 512);
  if ("stack" in obj) trimmed.stack = clipAny(obj.stack, 2048);
  return JSON.stringify(trimmed);
}

/** 校验单条事件；失败时返回 ItemError，不阻断整包。 */
export function validateOne(
  raw: RawEvent,
): { fields: EventFields; error?: undefined } | { fields?: undefined; error: ItemError } {
  const id = (raw.event_id ?? "").trim();
  if (!id) {
    return {
      error: { event_id: raw.event_id ?? "", code: "INVALID_EVENT", message: "event_id required" },
    };
  }
  const typ = (raw.event_type ?? "").trim();
  if (!allowedTypes.has(typ)) {
    return { error: { event_id: id, code: "INVALID_TYPE", message: "unknown event_type" } };
  }
  const name = (raw.event_name ?? "").trim();
  if (!name) {
    return { error: { event_id: id, code: "INVALID_EVENT", message: "event_name required" } };
  }
  const platform = (raw.platform ?? "").trim();
  if (platform && platform !== "ios" && platform !== "android" && platform !== "web") {
    return { error: { event_id: id, code: "INVALID_EVENT", message: "invalid platform" } };
  }

  let propsJson: string;
  try {
    propsJson = normalizeProps(raw.props);
  } catch {
    return { error: { event_id: id, code: "INVALID_EVENT", message: "invalid props" } };
  }

  const durationMs =
    raw.duration_ms === undefined || raw.duration_ms === null ? null : Number(raw.duration_ms);
  const clientTsMs =
    raw.client_ts_ms === undefined || raw.client_ts_ms === null ? null : Number(raw.client_ts_ms);

  return {
    fields: {
      eventId: id,
      eventType: typ,
      eventName: name,
      appVersion: (raw.app_version ?? "").trim(),
      platform,
      deviceBrand: (raw.device_brand ?? "").trim(),
      sessionId: (raw.session_id ?? "").trim(),
      durationMs: Number.isFinite(durationMs as number) ? (durationMs as number) : null,
      propsJson,
      clientTsMs: Number.isFinite(clientTsMs as number) ? (clientTsMs as number) : null,
      studentNo: (raw.student_no ?? "").trim(),
    },
  };
}
